util/releasenotes.py

```diff
@@ -0,0 +1,112 @@
+"""
+Copyright (C) 2018 Ocean Observatories Initiative.
+"""
+
+import re
+import os
+
+class ReleaseNotes:
+    """
+    Python Object representation of markdown formatted release notes.
+    """
+
+    # default value for the Release Notes file
+    DEFAULT_FILE = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "../",
+        "RELEASE_NOTES.md")
+
+    # value for when the version cannot be determined
+    UNKNOWN_VERSION = "0.0.1-unknown"
+
+    # Matches the component name that should be the first header of the release notes
+    # for example: "# Stream Engine"
+    _COMPONENT_NAME_PATTERN = re.compile(r'#\s+(?P<name>.*)');
+
+    # Matches Patterns such as: "# Development Release 1.0.0 2018-11-01"
+    _VERSION_PATTERN = re.compile(r'#\s+(?P<descriptor>[\w\s]+)\s+(?P<version>\S+)\s+(?P<date>\S*)\s*')
+
+    # singleton instance to use so that file isn't double parsed
+    __instance = None
+
+    @staticmethod
+    def instance(file=DEFAULT_FILE):
+        if not ReleaseNotes.__instance:
+            ReleaseNotes.__instance = ReleaseNotes(file)
+        return ReleaseNotes.__instance
+
+    def __init__(self, file=DEFAULT_FILE):
+        self.file = file
+        self._contents = None
+        self._component_name = None
+        self._latest_descriptor = None
+        self._latest_version = None
+        self._latest_date = None
+        self._parsed = False
+
+    def contents(self):
+        """
+        :return: the contents of the file as a list of strings, or None if the contents couldn't
+        be read
+        """
+        if not self._contents:
+            with open(self.file, 'r') as file_handle:
+                self._contents = file_handle.readlines()
+        return self._contents
+
+    def _parse(self):
+        """
+        :return: True if the file has been parsed, False if something went wrong.
+        """
+        if not self._parsed and self.contents():
+            for line in self.contents():
+
+                # find component name
+                if not self._component_name:
+                    match = re.search(ReleaseNotes._COMPONENT_NAME_PATTERN, line)
+                    if match:
+                        self._component_name = match.group("name")
+                        continue
+
+                # find latest version
+                if not self._latest_version:
+                    match = re.search(ReleaseNotes._VERSION_PATTERN, line)
+                    if match:
+                        self._latest_version = match.group("version")
+                        self._latest_descriptor = match.group("descriptor")
+                        self._latest_date = match.group("date")
+                        continue
+
+                # parsed all we want to know right now, perform quick return
+                if self._component_name and self._latest_version:
```

Comment:
Check is unnecessary. If `_component_name` or `_lastest_version` are not set, the previous checks will continue the loop before this is reached. We cannot get to this line of code without it always being true.

Reply:
the checks are needed to break out of processing rest of the lines in the file.  right now all I care that the parser grab are just the component name and latest version, if I were to want to parse the rest of the file I would take out the check.  for example, if I put in the following code to demonstrate this:

                # parsed all we want to know right now, perform quick return
                if self._component_name and self._latest_version:
                    print("parsing complete")
                    break

                print("unhandled line: "+line)

when I run the program I get:

unhandled line: 

parsing complete

when I comment out the check:

                # parsed all we want to know right now, perform quick return
                #if self._component_name and self._latest_version:
                #    print("parsing complete")
                #    break

                print("unhandled line: "+line)

when I run the program now:

unhandled line: 

unhandled line: 

unhandled line: Issue #13063 - Parse the stream engine version from the release notes

unhandled line: - Read the version from the release notes at runtime, use this for output to netcdf

unhandled line:   also display version in log when starting up

unhandled line: 

unhandled line: Issue #13276 - Fix gunicorn compatibility issue with previous change

unhandled line: 

... etc

Reply:
My original code showed just the break, not the conditional. When you changed it here, you removed the conditional and the break. Put the break in by itself.

Reply:
the conditional is still needed to handle the line returns. if I just put break in there, it will grab the component, then the next line is a blank line which will fall through to the break statement which will break out of the loop before it has a chance to process the version information

Reply:
That is a negatory. Let's walk through it. The first two conditional both check to see if a variable has been set. In the case you just suggested `_latest_version` had not been set, but `_component_name` had been. In that case the `_latest_version` conditional will continue to get triggered until `_latest_version` is set, and it will continue to `continue`, never reaching code below.

Reply:
I think the misunderstanding here is rooted in the indentation level of the `continue` statements. In the latest code posted by John, the `continue` statements are inside the `if match` statement for both the `_component_name` and `_latest_version` checks. In Dan's proposed code, the `continue` statement is outside the `if match` block for the `_component_name` check and inside the `if match` block for the `_latest_version` check. I think Dan intended for both `continue` statements to be outside the `if match` blocks.

Given the `continue`s inside the `if match` blocks, as currently coded, the `if self._component_name and self._latest_version` check is needed because upon failure to match, the code does not `continue`. However, if both `continue`s were outside the `if match` blocks, then the `if self._component_name and self._latest_version` check becomes unnecessary. I prefer this latter approach, with the `continue`s outside the `if match` blocks and the `break` statement by itself.

Reply:
Quick note, as written currently, the `_component_name` is allowed to occur either before or after the `_latest_version`. As I proposed it, (with `continue` outside `if match` blocks) `_component_name` must occur before `_latest_version`. If we want to tolerate out of order, we need to use the current approach or equivalent.

Reply:
excellent feedback.  I had missed the indention of the continue.  the lack of {}'s in python throw me for loops sometimes.  I'll move both the continues outside of the if blocks.